src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -392,3 +422,155 @@ g.test('shader_stages_and_visibility')
+    const bglEntries0: GPUBindGroupLayoutEntry[] = [
+      { binding: 0, visibility: GPUShaderStage.FRAGMENT, type: 'sampled-texture' },
+      {
+        binding: 1,
+        visibility: GPUShaderStage.FRAGMENT,
+        type: bindingType,
+        storageTextureFormat: bindingTexFormat,
+      },
+    ];
+    const bgEntries0: GPUBindGroupEntry[] = [
+      { binding: 0, resource: sampledView },
+      { binding: 1, resource: sampledStorageView },
+    ];
+    const bindGroup0 = t.device.createBindGroup({
+      entries: bgEntries0,
+      layout: t.device.createBindGroupLayout({ entries: bglEntries0 }),
+    });
+
+    // Create bindGroup1. It has one binding, which use the same view/subresoure of a binding in
+    // bindGroup0. So it may or may not conflicts with that binding in bindGroup0.
+    const bindGroup1 = t.createBindGroup(0, sampledStorageView, 'sampled-texture', undefined);
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: t.createTexture().createView(),
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+
+    // Set bindGroup0 and bindGroup1. bindGroup0 is replaced by bindGroup1 in the current pass.
+    // But bindings in bindGroup0 should be tracked too.
+    pass.setBindGroup(0, bindGroup0);
+    pass.setBindGroup(0, bindGroup1);
+    pass.endPass();
+
+    const success = bindingType === 'writeonly-storage-texture' ? false : true;
+    t.expectValidationError(() => {
+      encoder.finish();
+    }, !success);
+  });
+
+g.test('bindings_in_bundle')
+  .params(
+    params()
+      .combine(pbool('binding0InBundle'))
+      .combine(pbool('binding1InBundle'))
+      .combine(poptions('type0', ['render-target', ...kTextureBindingTypes] as const))
+      .combine(poptions('type1', ['render-target', ...kTextureBindingTypes] as const))
+      .unless(
+        ({ binding0InBundle, binding1InBundle, type0, type1 }) =>
+          // We can't set 'render-target' in bundle, so we need to exclude it from bundle.
+          // In addition, if both bindings are non-bundle, there is no need to test it because
+          // we have far more comprehensive test cases for that situation in this file.
+          (binding0InBundle && type0 === 'render-target') ||
+          (binding1InBundle && type1 === 'render-target') ||
+          (!binding0InBundle && !binding1InBundle)
+      )
+  )
+  .fn(async t => {
+    const { binding0InBundle, binding1InBundle, type0, type1 } = t.params;
+
+    // Two bindings are attached to the same texture view.
+    const view = t
+      .createTexture({
+        usage:
+          GPUTextureUsage.OUTPUT_ATTACHMENT | GPUTextureUsage.STORAGE | GPUTextureUsage.SAMPLED,
+      })
+      .createView();
+
+    const bindGroups: GPUBindGroup[] = [];
```

Comment:
uNit:
```suggestion
    const bindGroups: (GPUBindGroup | undefined)[] = [];
```

Reply:
Hmm. I tried this. But if failed when I run "npx grunt pre". The error reported that when we call setBindGroup(i, bindGroups[i]); type 'GPUBindGroup | undefined' is not assignable to parameter of type 'GPUBindGroup'.